lib/nostrum/cache/guild_cache.ex

```diff
@@ -24,6 +24,7 @@ defmodule Nostrum.Cache.GuildCache do
   @type reason ::
           :id_not_found
           | :id_not_found_on_guild_lookup
+          | :channel_not_found
```

Comment:
this can happen because `GuildCache.select_by` accepts a selector function (which can return anything) and passes on the result as-is
https://github.com/Kraigie/nostrum/blob/6c99bb8af945d8ba5dd22058baaf090c06132088/lib/nostrum/cache/guild_cache.ex#L22
https://github.com/Kraigie/nostrum/blob/6c99bb8af945d8ba5dd22058baaf090c06132088/lib/nostrum/cache/guild_cache.ex#L146
The specific place it happens is here:
https://github.com/Kraigie/nostrum/blob/603da771b0171c71e49109dfb72f26a912037380/lib/nostrum/cache/channel_cache.ex#L96-L98
An alternate (and maybe more correct) fix would be to make a new `@type selector_return`, modify `selector` to require returning that type, and allow `select_by` to return that as well

Reply:
I don't think your idea is bad but I have another idea I want to run by you.

Why don't we instead move the return out of the anonymous function. This way we're not introducing another potential return value of `{:error, :channel_not_found}` which we have to document in `GuildCache.select_by/2`